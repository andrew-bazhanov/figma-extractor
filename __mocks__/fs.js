import { fs } from 'memfs';

fs.__originalFs = jest.requireActual('fs');
module.exports = fs;
