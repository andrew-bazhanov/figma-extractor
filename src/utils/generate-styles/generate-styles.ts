import { FileNodesResponse, FullStyleMetadata, StyleType } from 'figma-js';
import fs from 'fs';

import { getColorStyles } from '../../lib/get-color-styles';
import { getEffectStyles } from '../../lib/get-effect-styles';
import { getGradientStyles } from '../../lib/get-gradient-styles';
import { getTextStyles } from '../../lib/get-text-styles';
import { stringifyRecordsWithSort } from '../../lib/stringify';
import { writeStyleFile } from '../../lib/write-style-file';

// eslint-disable-next-line @typescript-eslint/naming-convention
type StyleTypePredicate = (item: { style_type: StyleType }) => boolean;

const getStyleTypePredicate = (styleType: StyleType): StyleTypePredicate => {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  return ({ style_type }) => style_type === styleType;
};

const isStyleTypeFill = getStyleTypePredicate('FILL');
const isStyleTypeEffect = getStyleTypePredicate('EFFECT');
const isStyleTypeText = getStyleTypePredicate('TEXT');

export const generateStyles = (
  config: Config,
  styleMetadata: readonly FullStyleMetadata[],
  fileNodes: FileNodesResponse,
) => {
  if (!fs.existsSync(config?.styles?.exportPath)) {
    fs.mkdirSync(config?.styles?.exportPath, { recursive: true });
  }

  const metaColors = styleMetadata.filter(isStyleTypeFill);

  if (!config?.styles?.colors?.disabled) {
    const colors = getColorStyles(metaColors, fileNodes, config);
    const colorTemplate = `module.exports = ${stringifyRecordsWithSort(colors)} ;`;
    writeStyleFile(colorTemplate, 'colors.js', config);
  }

  if (!config?.styles?.gradients?.disabled) {
    const gradients = getGradientStyles(metaColors, fileNodes, config);
    const gradientsTemplate = `module.exports = ${stringifyRecordsWithSort(gradients)} ;`;
    writeStyleFile(gradientsTemplate, 'gradients.js', config);
  }

  if (!config?.styles?.effects?.disabled) {
    const metaEffects = styleMetadata.filter(isStyleTypeEffect);
    const effects = getEffectStyles(metaEffects, fileNodes, config);
    const effectTemplate = `module.exports = {boxShadow: ${stringifyRecordsWithSort(effects)}} ;`;
    writeStyleFile(effectTemplate, 'effects.js', config);
  }

  if (!config?.styles?.textStyles?.disabled) {
    const metaTextStyles = styleMetadata.filter(isStyleTypeText);
    const { fontFamily, textStyles } = getTextStyles(metaTextStyles, fileNodes, config);
    const fontFamilyTemplate = JSON.stringify(fontFamily);
    const textStylesTemplate = `{${textStyles.join()}};`;
    writeStyleFile(
      `
      const fontFamily = ${fontFamilyTemplate};
      const textVariants = ${textStylesTemplate};
      module.exports = {fontFamily, textVariants}`,
      'text-styles.js',
      config,
    );
  }

  return false;
};
