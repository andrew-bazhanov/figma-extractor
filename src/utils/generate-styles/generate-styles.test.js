import fs from 'fs';
import { vol } from 'memfs';

import { fileNodes } from './fixtures/file-nodes';
import { styleMetadata } from './fixtures/style-metadata';
import { generateStyles } from './generate-styles';

jest.mock('fs');

// The default config has all style items disabled
const getConfig = ({
  disableColors = true,
  disableGradients = true,
  disableEffects = true,
  disableTextStyles = true,
  disableIcons = false,
} = {}) => ({
  apiKey: 'dummy-api-key',
  fileId: 'dummy-file-id',
  styles: {
    exportPath: '/ui/theme',
    colors: {
      disabled: disableColors,
    },
    gradients: {
      disabled: disableGradients,
    },
    effects: {
      disabled: disableEffects,
    },
    textStyles: {
      disabled: disableTextStyles,
    },
  },
  icons: {
    disabled: disableIcons,
  },
});

describe('generateStyles', () => {
  afterEach(() => {
    vol.reset();
  });

  it('creates styles dir according to provided config if there is none', () => {
    const config = getConfig();

    expect(fs.existsSync('/ui/theme')).toBe(false);

    generateStyles(config, styleMetadata.styles, fileNodes);

    expect(fs.existsSync('/ui/theme')).toBe(true);
  });

  it('creates file with colors according to provided config', () => {
    const config = getConfig({
      disableColors: false,
    });

    // eslint-disable-next-line @typescript-eslint/no-var-requires
    generateStyles(config, styleMetadata.styles, fileNodes);

    expect(vol.toJSON()).toMatchSnapshot();
  });

  it('creates file with effects according to provided config', () => {
    const config = getConfig({
      disableEffects: false,
    });

    // eslint-disable-next-line @typescript-eslint/no-var-requires
    generateStyles(config, styleMetadata.styles, fileNodes);

    expect(vol.toJSON()).toMatchSnapshot();
  });

  it('creates file with text styles according to provided config', () => {
    const config = getConfig({
      disableTextStyles: false,
    });

    // eslint-disable-next-line @typescript-eslint/no-var-requires
    generateStyles(config, styleMetadata.styles, fileNodes);

    expect(vol.toJSON()).toMatchSnapshot();
  });
});
