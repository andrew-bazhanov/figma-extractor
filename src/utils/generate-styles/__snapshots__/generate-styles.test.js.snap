// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateStyles creates file with colors according to provided config 1`] = `
Object {
  "/ui/theme/colors.js": "module.exports = {\\"text/txt900\\":\\"#2a3fff\\"} ;",
}
`;

exports[`generateStyles creates file with effects according to provided config 1`] = `
Object {
  "/ui/theme/effects.js": "module.exports = {boxShadow: {\\"shadow200\\":\\"0px 18px 88px rgba(24, 39, 75, 0.14)\\"}} ;",
}
`;

exports[`generateStyles creates file with text styles according to provided config 1`] = `
Object {
  "/ui/theme/text-styles.js": "
      const fontFamily = {\\"font1\\":\\"'Hind Vadodara', Arial, sans-serif\\"};
      const textVariants = {'heading/bs-h500': {
      fontFamily: fontFamily.font1,
      fontSize: 32,
      fontWeight: 600,
      textTransform: \\"initial\\",
      lineHeight: 1.12,
    }};;
      module.exports = {fontFamily, textVariants}",
}
`;
